lib/src/merged_tree.rs

```diff
@@ -659,56 +684,8 @@ impl<'matcher> TreeDiffIterator<'matcher> {
             Ok(Merge::resolved(Tree::null(store.clone(), dir.to_owned())))
         }
     }
-}
 
-impl TreeDiffDirItem {
-    fn from_trees(
-        dir: &RepoPath,
-        trees1: &Merge<Tree>,
-        trees2: &Merge<Tree>,
-        matcher: &dyn Matcher,
-    ) -> Self {
-        let mut entries = vec![];
-        for (name, before, after) in merged_tree_entry_diff(trees1, trees2) {
-            let path = dir.join(name);
-            let before = before.to_merge();
-            let after = after.to_merge();
-            let tree_before = before.is_tree();
-            let tree_after = after.is_tree();
-            // Check if trees and files match, but only if either side is a tree or a file
-            // (don't query the matcher unnecessarily).
-            let tree_matches = (tree_before || tree_after) && !matcher.visit(&path).is_nothing();
-            let file_matches = (!tree_before || !tree_after) && matcher.matches(&path);
-
-            // Replace trees or files that don't match by `Merge::absent()`
-            let before = if (tree_before && tree_matches) || (!tree_before && file_matches) {
-                before
-            } else {
-                Merge::absent()
-            };
-            let after = if (tree_after && tree_matches) || (!tree_after && file_matches) {
-                after
-            } else {
-                Merge::absent()
-            };
-            if before.is_absent() && after.is_absent() {
-                continue;
-            }
-            entries.push(TreeDiffEntry {
-                source: path.clone(),
-                target: path,
-                value: Ok((before, after)),
-            });
-        }
-        entries.reverse();
-        Self { entries }
-    }
-}
-
-impl Iterator for TreeDiffIterator<'_> {
-    type Item = TreeDiffEntry;
-
-    fn next(&mut self) -> Option<Self::Item> {
+    fn next_impl(&mut self) -> Option<TreeDiffEntry> {
```

Comment:
nit: could you do the refactoring of defining `next_impl()` in a separate commit? thanks

Reply:
done